midterm-2023-11-06/code/5115-nikola-tosic/script.ts

```diff
@@ -0,0 +1,162 @@
+interface Album {
+    name: string;
+    year: number;
+}
+
+interface Band {
+    id: number;
+    name: string;
+    formed: number;
+    location: string;
+    country: string;
+    members: string[];
+    albums: Album[];
+}
+
+type BandSorter = (first: Band, second: Band) => number;
+
+document.addEventListener("DOMContentLoaded", siteCode);
+
+let bands: Band[] = [];
+
+async function siteCode() {
+    const data = await loadData();
+    bands = data;
+
+    fillCountries(bands);
+
+    displayBands(bands);
+
+    const nameSort = document.getElementById("sort-name")!;
+    nameSort.addEventListener("click", sortByName);
+
+    const idSort = document.getElementById("sort-id")!;
+    idSort.addEventListener("click", sortById);
+
+    const applyFilterButton = document.getElementById("apply-filter")!;
+    applyFilterButton.addEventListener("click", applyFilter);
+
+    const modal = document.getElementById("album-details")!;
+    modal.addEventListener("click", () => {
+        modal.classList.add("hidden");
+    });
+}
+
+const nameSorter: BandSorter = (first, second) => first.name.localeCompare(second.name);
+const idSorter: BandSorter = (first, second) => first.id - second.id;
+
+const sortByName = () => {
+    const sortedBands = bands.slice().sort(nameSorter);
+    displayBands(sortedBands);
+};
+
+const sortById = () => {
+    const sortedBands = bands.slice().sort(idSorter);
+    displayBands(sortedBands);
+};
+
+const fillCountries = (bands: Band[]) => {
+    const filter = document.getElementById("country-filter")!;
+
+    const countries = new Set<string>();
+    for (const band of bands) {
+        countries.add(band.country);
+    }
+
+    for (const country of countries) {
+        const option = document.createElement("option");
+        option.value = country;
+        option.innerHTML = country;
+        filter.appendChild(option);
+    }
+};
+
+const applyFilter = () => {
+    const countryElement = document.getElementById("country-filter") as HTMLSelectElement;
+    const country = countryElement.value;
+
+    let filteredBands = bands;
+    if (country !== "all") {
+        filteredBands = filteredBands.filter(band => band.country === country);
+    }
+
+    displayBands(filteredBands);
+};
+
+const loadData = async () => {
+    const dataUri = "https://raw.githubusercontent.com/sweko/uacs-internet-programming-exams/main/midterm-2023-11-06/data/bands.json";
+    const response = await fetch(dataUri);
+
+    if (!response.ok) {
+        throw new Error("Failed to fetch data");
+    }
+
+    const data = await response.json();
+    return data.metalBands;
+};
+
+const displayBands = (bands: Band[]) => {
+    const container = document.getElementById("band-container")!;
+    container.innerHTML = "";
+    for (const band of bands) {
+        const bandRow = generateBandRow(band);
+        container.appendChild(bandRow);
+    }
+};
+
+const generateBandRow = (band: Band) => {
```

Comment:
Some issues with displaying the data, since the row does not have the `band-table` class, the data is displayed badly.  All the data is visible, though, so  +10 points.